File: java-sdk/src/main/java/com/framework/connection/ConnectionLifecycleManager.java
package com.framework.connection;

import com.framework.exception.ErrorCode;
import com.framework.exception.FrameworkException;
import com.framework.protocol.router.ServiceEndpoint;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * 连接生命周期管理器
 * 
 * 负责连接的空闲超时关闭、失败重连和优雅关闭
 * 包装 DefaultConnectionManager，增加重连和生命周期管理能力
 * 
 * **验证需求: 7.2, 7.3, 7.5**
 */
public class ConnectionLifecycleManager {

    private static final Logger logger = LoggerFactory.getLogger(ConnectionLifecycleManager.class);

    private final DefaultConnectionManager connectionManager;
    private final ConnectionConfig config;
    private final AtomicBoolean closed;

    /** 重连任务调度器 */
    private final ScheduledExecutorService reconnectScheduler;

    /** 正在重连的端点集合 */
    private final ConcurrentHashMap<String, AtomicBoolean> reconnecting;

    public ConnectionLifecycleManager(DefaultConnectionManager connectionManager,
                                      ConnectionConfig config) {
        this.connectionManager = connectionManager;
        this.config = config;
        this.closed = new AtomicBoolean(false);
        this.reconnecting = new ConcurrentHashMap<>();

        this.reconnectScheduler = Executors.newScheduledThreadPool(2, r -> {
            Thread t = new Thread(r, "conn-reconnect");
            t.setDaemon(true);
            return t;
        });
    }

    /**
     * 获取连接，失败时自动重连
     */
    public ManagedConnection getConnectionWithRetry(ServiceEndpoint endpoint) {
        if (closed.get()) {
            throw new FrameworkException(ErrorCode.CONNECTION_ERROR, "生命周期管理器已关闭");
        }

        int attempts = 0;
        long delay = config.getReconnectDelayMs();
        FrameworkException lastException = null;

        while (attempts <= config.getMaxReconnectAttempts()) {
            try {
                return connectionManager.getConnection(endpoint);
            } catch (FrameworkException e) {
                lastException = e;
                if (e.getErrorCode() != ErrorCode.CONNECTION_ERROR) {
                    throw e; // 非连接错误直接抛出
                }

                attempts++;
                if (attempts > config.getMaxReconnectAttempts()) {
                    break;
                }

                logger.warn("连接失败, 第 {}/{} 次重试, endpoint={}, delay={}ms",
                        attempts, config.getMaxReconnectAttempts(), endpoint, delay);

                try {
                    Thread.sleep(delay);
                } catch (InterruptedException ie) {
                    Thread.currentThread().interrupt();
                    throw new FrameworkException(ErrorCode.CONNECTION_ERROR,
                            "重连被中断", ie);
                }

                // 指数退避
                delay = Math.min(delay * 2, 30_000);
            }
        }

        throw new FrameworkException(ErrorCode.CONNECTION_ERROR,
                "重连失败: 已达最大重试次数 " + config.getMaxReconnectAttempts()
                        + ", endpoint=" + endpoint,
                lastException);
    }

    /**
     * 异步重连到指定端点
     * 使用指数退避策略
     */
    public CompletableFuture<ManagedConnection> reconnectAsync(ServiceEndpoint endpoint) {
        String key = endpoint.getAddress() + ":" + endpoint.getPort();
        AtomicBoolean inProgress = reconnecting.computeIfAbsent(key, k -> new AtomicBoolean(false));

        if (!inProgress.compareAndSet(false, true)) {
            // 已有重连任务在进行
            return CompletableFuture.failedFuture(
                    new FrameworkException(ErrorCode.CONNECTION_ERROR,
                            "重连已在进行中: " + endpoint));
        }

        CompletableFuture<ManagedConnection> future = new CompletableFuture<>();

        reconnectScheduler.submit(() -> {
            try {
                ManagedConnection conn = getConnectionWithRetry(endpoint);
                future.complete(conn);
            } catch (Exception e) {
                future.completeExceptionally(e);
            } finally {
                inProgress.set(false);
            }
        });

        return future;
    }

    /**
     * 释放连接
     */
    public void releaseConnection(ManagedConnection connection) {
        connectionManager.releaseConnection(connection);
    }

    /**
     * 优雅关闭
     * 等待所有活跃请求完成后关闭所有连接
     */
    public CompletableFuture<Void> shutdownGracefully(long timeoutMs) {
        if (!closed.compareAndSet(false, true)) {
            return CompletableFuture.completedFuture(null);
        }

        logger.info("优雅关闭连接生命周期管理器, 超时={}ms", timeoutMs);

        reconnectScheduler.shutdown();
        try {
            if (!reconnectScheduler.awaitTermination(timeoutMs, TimeUnit.MILLISECONDS)) {
                reconnectScheduler.shutdownNow();
            }
        } catch (InterruptedException e) {
            reconnectScheduler.shutdownNow();
            Thread.currentThread().interrupt();
        }

        return connectionManager.shutdownGracefully(timeoutMs);
    }

    /**
     * 强制关闭
     */
    public CompletableFuture<Void> close() {
        if (!closed.compareAndSet(false, true)) {
            return CompletableFuture.completedFuture(null);
        }

        reconnectScheduler.shutdownNow();
        return connectionManager.closeAll();
    }

    /**
     * 获取连接池统计
     */
    public ConnectionPoolStats getPoolStats(ServiceEndpoint endpoint) {
        return connectionManager.getPoolStats(endpoint);
    }

    /**
     * 获取全局统计
     */
    public ConnectionPoolStats getTotalStats() {
        return connectionManager.getTotalStats();
    }

    public boolean isClosed() {
        return closed.get();
    }

    public DefaultConnectionManager getConnectionManager() {
        return connectionManager;
    }
}

File: golang-sdk/connection/connection_lifecycle_manager.go
package connection

import (
	"context"
	"fmt"
	"sync"
	"sync/atomic"
	"time"
)

// ConnectionLifecycleManager 连接生命周期管理器
//
// 包装 DefaultConnectionManager，增加重连和生命周期管理能力
//
// **验证需求: 7.2, 7.3, 7.5**
type ConnectionLifecycleManager struct {
	manager      ConnectionManager
	config       *ConnectionConfig
	closed       atomic.Bool
	reconnecting sync.Map // map[string]*atomic.Bool
}

// NewConnectionLifecycleManager 创建连接生命周期管理器
func NewConnectionLifecycleManager(manager ConnectionManager, config *ConnectionConfig) *ConnectionLifecycleManager {
	if config == nil {
		config = DefaultConnectionConfig()
	}

	return &ConnectionLifecycleManager{
		manager: manager,
		config:  config,
	}
}

// GetConnectionWithRetry 获取连接（带重试）
//
// 如果连接失败，会根据配置的重连策略进行重试
//
// **验证需求: 7.3**
func (m *ConnectionLifecycleManager) GetConnectionWithRetry(ctx context.Context, endpoint *ServiceEndpoint) (*ManagedConnection, error) {
	if m.closed.Load() {
		return nil, fmt.Errorf("connection lifecycle manager is closed")
	}

	key := endpointKey(endpoint)

	// 检查是否正在重连
	if reconnecting, ok := m.reconnecting.Load(key); ok {
		if reconnecting.(*atomic.Bool).Load() {
			return nil, fmt.Errorf("endpoint %s is reconnecting", key)
		}
	}

	var lastErr error
	attempts := 0
	maxAttempts := m.config.MaxReconnectAttempts

	for attempts <= maxAttempts {
		attempts++

		// 尝试获取连接
		conn, err := m.manager.GetConnection(ctx, endpoint)
		if err == nil {
			return conn, nil
		}

		lastErr = err

		// 如果是最后一次尝试，不再等待
		if attempts > maxAttempts {
			break
		}

		// 计算退避延迟（指数退避）
		delay := m.config.ReconnectDelay * time.Duration(1<<uint(attempts-1))
		if delay > 30*time.Second {
			delay = 30 * time.Second
		}

		// 等待后重试
		select {
		case <-ctx.Done():
			return nil, ctx.Err()
		case <-time.After(delay):
			// 继续重试
		}
	}

	return nil, fmt.Errorf("failed to connect after %d attempts: %w", attempts, lastErr)
}

// ReconnectAsync 异步重连到指定端点
//
// 在后台尝试重新建立连接
//
// **验证需求: 7.3**
func (m *ConnectionLifecycleManager) ReconnectAsync(endpoint *ServiceEndpoint) {
	key := endpointKey(endpoint)

	// 创建或获取重连标志
	reconnectingInterface, _ := m.reconnecting.LoadOrStore(key, &atomic.Bool{})
	reconnecting := reconnectingInterface.(*atomic.Bool)

	// 如果已经在重连，直接返回
	if !reconnecting.CompareAndSwap(false, true) {
		return
	}

	go func() {
		defer reconnecting.Store(false)

		ctx := context.Background()
		_, _ = m.GetConnectionWithRetry(ctx, endpoint)
	}()
}

// ReleaseConnection 释放连接
func (m *ConnectionLifecycleManager) ReleaseConnection(conn *ManagedConnection) {
	m.manager.ReleaseConnection(conn)
}

// CloseConnections 关闭到指定端点的所有连接
func (m *ConnectionLifecycleManager) CloseConnections(endpoint *ServiceEndpoint) error {
	return m.manager.CloseConnections(endpoint)
}

// ShutdownGracefully 优雅关闭
//
// 等待所有活跃连接完成后关闭
//
// **验证需求: 7.5**
func (m *ConnectionLifecycleManager) ShutdownGracefully(timeout time.Duration) error {
	if !m.closed.CompareAndSwap(false, true) {
		return nil
	}

	return m.manager.ShutdownGracefully(timeout)
}

// CloseAll 关闭所有连接
func (m *ConnectionLifecycleManager) CloseAll() error {
	if !m.closed.CompareAndSwap(false, true) {
		return nil
	}

	return m.manager.CloseAll()
}

// GetPoolStats 获取连接池统计
func (m *ConnectionLifecycleManager) GetPoolStats(endpoint *ServiceEndpoint) *ConnectionPoolStats {
	return m.manager.GetPoolStats(endpoint)
}

// GetTotalStats 获取全局统计
func (m *ConnectionLifecycleManager) GetTotalStats() *ConnectionPoolStats {
	return m.manager.GetTotalStats()
}

// UpdateConfig 更新配置
func (m *ConnectionLifecycleManager) UpdateConfig(config *ConnectionConfig) {
	m.config = config
	m.manager.UpdateConfig(config)
}

// IsClosed 检查是否已关闭
func (m *ConnectionLifecycleManager) IsClosed() bool {
	return m.closed.Load()
}

// GetManager 获取底层连接管理器
func (m *ConnectionLifecycleManager) GetManager() ConnectionManager {
	return m.manager
}
